### ISSUE TITLE
New WP calculator - Get rid of placeholder variables

### ISSUE THREAD

**Issue by Lineflyer** [cgeo/cgeo#12405]
While trying out the new waypoint calculator, which is combined with the cache variable tab, I had somehow a hard time to understand how its used.

@eddiemuc 
Can we talk through the following example here:

Typical multi cache:

- Stage 1: You will find the values for x and y at this stage.
- Stage 2: Go to N 51° 21.(x+1)y5 E006° 3x.650. You will find value z there.
- Final: Go to N 51° 21.z(y+2)x E006° 3x.(z+3)y3

With the old waypoint calculator I need to fill the calculator for stage 2 and final as follows:

- Waypoint for Stage 2:  N51° 21.AB5 E006° 3C.650  with A=x+1, B=y, C=x
- Waypoint for Final: N51° 21.ABC E006° 3D.EF3 with A=z, B=y+2, C=x, D=x, E=z+3, F=y

The drawback of these non global variables (lower case ones) are now, that I have to enter the values for x,y,z on each waypoint (where its required).

I would have expected, that the new approach stays the same with (in principle) the only difference, that the values of those lower case variables are now retrieved from the result/value in the variable tab. 
But either I don`t understand how to achieve this (which would put into question if an average user will understand the changed working mode) or its implemented in a different way (in which case I would ask why).

**Comment by Lineflyer** [comment]
I made a user defined cache containing the example above to visualize the current mode of operation I use:

Stage 2 from above example:
![Screenshot_20211225-180002_cgeo](https://user-images.githubusercontent.com/949669/147389897-5c414575-21ad-446d-b0e9-5c971556f35a.jpg)

Final from above example:
![Screenshot_20211225-180111_cgeo](https://user-images.githubusercontent.com/949669/147389908-e9a14302-b45b-4601-84e3-b39313527707.jpg)

IMHO the "only" thing needed is to "remove" the right input side (lower case variables) and retrieve them from the variable tab instead, or not?

When trying the same in the new WP calculator I get much more than needed on the variable tab:
![Screenshot_20211225-180535_cgeo](https://user-images.githubusercontent.com/949669/147389987-d4870ee1-eeb4-474a-a390-626a69652149.jpg)
![Screenshot_20211225-180557_cgeo](https://user-images.githubusercontent.com/949669/147389989-b661a0a8-b087-4e95-a509-6fe0f520b3cb.jpg)

**Comment by Lineflyer** [comment]
Additionally I am now totally stuck when trying to create the final waypoint in the new calculator as the auto-created A, B, C when tapping on the digits are already prefilled with what I entered for the stage2 waypoint.

Unless I am not understanding the working mode, I suggest to keep the IMHO easier and more understandable upper-case/lower-case distinction (or a clever alternative) to differentiate between WP-local variables (upper case) and cache global variables (lower-case).

**Comment by eddiemuc** [comment]
@Lineflyer I think you are a bit stuck in the unnecessary complexity which was introduced with the "old" model since you somehow still try to differentiate between "uppercase" and "lowercase" and unnecessarily introduce "inbetween" vars.

 In the new working model there is no differentiation between "global" and "local" variables any more, all variables are "global". Also there is no differentiation between "uppercase" and "lowercase" variables any more, all are handled the same and it is up to the user to decide whether he wants to work rather with uppercase variables, lowercase variables or a mixture of both.

In your example, you unnecessarily introduce variables A, B, C, D, E, F where you would only need x, y and z.

Let me show you how your example is supposed to be handled in the new model.

(I filled in values X=1, Y=2, Z=3 for illustration purposes)

Waypoint for Stage 2:
![image](https://user-images.githubusercontent.com/6909759/147390139-4ef33753-9948-45d7-a5c5-86cff788d17c.png)

Waypoint for Final:
![image](https://user-images.githubusercontent.com/6909759/147390160-6ce43270-d768-4ed3-aad9-5d1a815203f8.png)

Variable Tab:
![image](https://user-images.githubusercontent.com/6909759/147390164-4f062ef1-3fd0-4e0f-85c5-c4f0134ced48.png)

And that's all. Why would you need variables A, B, C, D, E or F?

Now I noticed that you used the strange "button-per-digit" interface whose purpose I never understood (I only use plain input of course). This "button-per-digit" interface in its current form allows you to enter X, Y, Z only by long-pressing each single digit. You might use A, B,C, D, E, F as intermediate variables instead if you want, but you end up of course with the self-induced complexity you described above - because A-F are global variables too.

As mentioned before, I would be all in favor to drop the "button-per-digit" interface, I don't see and never saw a real use for it. Alternatively, we may change the rules by which variables are introduced there on click. What are your thoughts on this?

**Comment by eddiemuc** [comment]
> Additionally I am now totally stuck when trying to create the final waypoint in the new calculator as the auto-created A, B, C when tapping on the digits are already prefilled with what I entered for the stage2 waypoint.

Unless we drop the "button-per-digit" view altogether, this specific aspect can be solved by introducing a global new variable on each click. E.g. if variables A, B, C already exist for the cache, then a click on a digit introduced the next unused one-digit-variable "D". Clicking on another digit would then introduce "E" and so on. Would that be a solution?

**Comment by Lineflyer** [comment]
> @Lineflyer I think you are a bit stuck in the unnecessary complexity which was introduced with the "old" model since you somehow still try to differentiate between "uppercase" and "lowercase" and unnecessarily introduce "inbetween" vars.

Indeed. Thanks for explaining, now I got what the aim is. I ever only used the button-mode and probably you started from the plain mode where this all makes more sense then.

> Unless we drop the "button-per-digit" view altogether, this specific aspect can be solved by introducing a global new variable on each click. E.g. if variables A, B, C already exist for the cache, then a click on a digit introduced the next unused one-digit-variable "D". Clicking on another digit would then introduce "E" and so on. Would that be a solution?

I would like to avoid dropping the "button-per-digit" mode as its kind of easier to fill out for many users.
How about still using it as guided input but not assigning any variable but - as soon as a digit is clicked - open a dialog where the user can enter the formula/variable for that digit?

Example: 
I want to have N51° 12.(x+9)45, so coming from N 51° 12.345 a click on "3" opens a dialog and lets the user enter x+9 

The difficult part would be how to visualize this in the best way, that's a real challenge.
I could imagine to start with the plain view as default and have a button "Guided input" which opens the "button mode" allows assigning formulas/variables (as just described above) and once you confirm at the end, you are back to the plain view now containing what you entered before during the guided input. Can you follow, what I mean?

**Comment by eddiemuc** [comment]
I think I can follow.

Unlike before, the new button-view is now coded simply as a "fancy front" to enter a plain formula anyway, so in principal what you suggest is easy to do. The hard party as you say, is how to best visualize it. 
Maybe we have to let go of the idea that each "digit button" has the same width and that the buttons get wider if more text is entered (such as `x+9` in your example). Don't know how this will look though

**Comment by Lineflyer** [comment]
Not sure if its really needed to have a direct visualization. Maybe the digits you replaced with a variable/formula in such case would just be replaced by a green tick mark in that digit and once you leave the "guided input" you would only see the result in the plain view again.

Mock up:
After having done your inputs:
![image](https://user-images.githubusercontent.com/949669/147391799-0eb06002-4691-4a3b-a5d6-cc17107fd33e.png)

When you leave (back to plain) it would show what you entered:
![image](https://user-images.githubusercontent.com/949669/147391824-cdc8f7fc-fd1c-4e09-bd56-039d6f55d418.png)

**Comment by Lineflyer** [comment]
Renamed the issue to focus on the retirement of that placeholder variables. IMHO thats the key point to make the whole interworking understandable.

**Comment by eddiemuc** [comment]
> Maybe the digits you replaced with a variable/formula in such case would just be replaced by a green tick mark in that digit 

In your example example, the user would see the green ticks, but under it the variables X and Y would show (because the are used in the coordinates). The user at this point could not see why x and y are shown (he may remember that he just entered them, but if he saves at this point and opens coordinate input later this might be puzzling to him)

**Comment by Lineflyer** [comment]
I was talking about a dedicated windows for this button mode. No need to show the lower part of the current screen on this one.
I played around in paint to visualize it:

Step 1: 
The initial view on the WP calculator screen would be only plain (note: no more selection for input mode on top).
It could/should IMHO be prefilled with current cords or cache coords.
There is now a new button "Guide me" (or whatever text).
![flow_1](https://user-images.githubusercontent.com/949669/147392772-e67f1bc2-8852-4f8c-b7df-ced6ebb2ff50.jpg)

Step 2:
When using "Guide me" a new dialog opens with the button mode (prefilled from the plain mode in step 1) and a text box below which is active when tapping on a digit.
The user can either enter a plain number or a formula/variable.
Remark: What was the long click before would now be the short click (old short click to fill with placeholder no longer needed).
![flow_2](https://user-images.githubusercontent.com/949669/147392821-79225565-5eb1-427a-9520-64126c88f19d.jpg)
Once done there should either be an OK button (not shown above) or the user confirms by tapping on the next digit.

Step 3:
Repeat for next digit(s).
Digits filled with a number will show the number, those filled with variable can show the variable, those filled with a formula should show another generic placeholder (green tick here, but maybe better a formula symbol(?)).
![flow_3](https://user-images.githubusercontent.com/949669/147392867-41ec590e-7733-4237-877e-ac073e222e6f.jpg)

Step 4:
Once you entered everything you need for this WP select "Fertig" to return to the main view from step 1.
![flow_4](https://user-images.githubusercontent.com/949669/147392878-30d27c87-c485-4ada-8bcc-c5978041c6d8.jpg)

Step 5:
Back at the view from step 1 you will now see the result in plain mode and the variable input below.
![flow_5](https://user-images.githubusercontent.com/949669/147392892-5b1c232d-a0f6-486a-9a27-2661644249cf.jpg)

**Comment by Lineflyer** [comment]
Just noticed, that my steps above do not show consistence in the values (step 5)...but I guess you still get what I mean.

**Comment by Lineflyer** [comment]
Remark regarding step 1:
> The initial view on the WP calculator screen would be only plain (note: no more selection for input mode on top).

Best would probably be to just reuse the normal WP coord input (you can still choose between differnt coord formats but button-mode is no longer there).

**Comment by Lineflyer** [comment]
@eddiemuc What do you think about the mocked up flow above?

**Comment by eddiemuc** [comment]
> @eddiemuc What do you think about the mocked up flow above?

Sorry didn't have the time yet to look into it. I will take a look today and reply later

**Comment by eddiemuc** [comment]
> @eddiemuc What do you think about the mocked up flow above?

Ok, some thoughts:
* If "guide me" is opened from "plain" mode, then there can be any kind of text input in the fields already. Might form a valid coordinate or not, might even be empty. Should the guide in these cases just be filled with current location? (This is what the old calculator does)
* The old calculator has the possibility to fill in a "space". Special handling needed for that?

**Comment by eddiemuc** [comment]
@Lineflyer while starting to implement a new dialog for the "guiding function" I thought about this issue again. I asked myself the question "why would users need such a guide?" I came up with the following main reasons:
* Cache coordinates are more nicely formatted and better readable because they are aligned vertically
* Cache coordinates (as long as they don't contain formulas yet) can be converted between different formats (DD.DDDD, DD MM.MMM etc) as a starting point for own editing.

COming from this I have an alternative suggestion to your proposal to approach this in a much less complex way:

* REMOVE the "format selection" (only PLAIN is available)
* Change the Typeface of the lat/lon input field to Monospace. This way it is automatically more aligned and better readable for edit (see below)
* Add a button "Prefill", when tapped the following list of options is displayed: "Current location in DD.DDDDD, Current location in DD MM.MMMM, current locaton in DD MM SS.SSSS, cache coordinates in DD.DDDDD, cc in DD MM.MMMM, current locaton in DD MM SS.SSSS) (can be formatted hierarchically)

See the following mockups:

![image](https://user-images.githubusercontent.com/6909759/147856529-984beb3a-2e4f-4be2-92f6-5e11c29400c3.png)

On tapping "prefill" some sort of selection list opens providing either cache or current location in three different formats:
![image](https://user-images.githubusercontent.com/6909759/147856569-97174e58-4470-4062-9535-15de4f6575cd.png)


Would that be a way to go? It would drastically reduce code complexity, we could get rid of several hundred lines of complex Java.

I honestly don't know which user group is using these "guided digit view" these days and what the reasons are for this. GCBuddy does not have this and I wouldn't use it myself. I want to find a way to both honor established usage patterns but reduce the functionality and GUI to what's really needed. 
If there is a dedicated user group for this I would really like to know which usage aspects of this GUI are important to them so we can preserve this.

**Comment by eddiemuc** [comment]
Updated mockup in previous comment (I striked through "Plain" spinner in red)

**Comment by Lineflyer** [comment]
Will read through it and provide feedback asap.

**Comment by eddiemuc** [comment]
@Lineflyer other idea: IF the fuided input IS needed, I propose not to put it in a separate dialog but to make it enabling using a sswitch "Guided" or something. Reason is that the user then has the chance also to see variables and the resulting calculation in parallel in the same screen. Something like this (mockup):

![image](https://user-images.githubusercontent.com/6909759/147872620-471e7637-66e3-44bc-92c9-96bfeeabe3f4.png)

Changes to ypur porposal:
* Integration in same dialog
* Switch in upper left "Guided". Selection between three coordinate formats to the right of it (only visible in Guided mode; In Plain mode I could place the "remove spaces" button there instead, this function would not be relevant in guided mode)
* Tapping a digit "marks" it visually (in my example I put an orange bar around it in the "lat" line)
* Field "Value/Formula" under "grid view" would only be enabled if a digit is marked. 
* As you already proposed: single letter/digit inputs would directly be shown in grid view. input with more than one char would show as something else instead (you used a green checkmark, I used a "..." in my mockup, tbd what to use finally)
* I would also assign the "next" action of the value/formula input field such that the next digit is marked

Again, **I do not need or personally see much use of such a guided input myself (neither today nor in the future)**. I am however willing to invest the 1-2 days effort to enable it again in the new model if there's a user community really in need for something like this. But it should then be designed in a way that it is really useful for this user group - and I am currently only guessing what it is they seek.

**Comment by Lineflyer** [comment]
I think a kind of guided input is very helpful for a lot of users to either get them onboarded into how to construct coordinate formulas with this function (maybe to later directly enter them) or also simply as a more convenient way to enter formulas for each digit without the risk of forgetting any paranthesis, etc.

I myself use the current waypoint calc in 99% of cases with this "digit-input" mode as I simply pretty much like the handling on a small mobile phone screen/keyboard. Usually I paste the owners instruction from the listing into the note field below and simply work along the digits to either enter the needed fixed value or construct the necessary calculation digit-by-digit.

That being said, looking at your proposals I definitely the second/latest idea!!!
It looks not only like a guided input but even outperforms the old waypoints calcs usability. Love it.
Its much better in guidance, than what you proposed at first and it puts less importance on the different formats (IMHO everything besides MinDec notation is quite rare used).

Feel free to implement as proposed. Once its in we can play around with it and finetune if needed.


Some comments to your remarks below:

> Integration in same dialog

Even better, so the user can see for which digit he currently enters a formula

> Switch in upper left "Guided". Selection between three coordinate formats to the right of it (only visible in Guided mode; In Plain mode I could place the "remove spaces" button there instead, this function would not be relevant in guided mode)

Sounds good (and looks good on your mock). So in plain mode, there is no need to define the format in beforehand? So how does c:geo then know whether its e.g. MinDec or DecDeg?

> Tapping a digit "marks" it visually (in my example I put an orange bar around it in the "lat" line)

As said above, I think its even an advantage to be able to see "where you are" while entering.

> Field "Value/Formula" under "grid view" would only be enabled if a digit is marked.

Logical.

> As you already proposed: single letter/digit inputs would directly be shown in grid view. input with more than one char would show as something else instead (you used a green checkmark, I used a "..." in my mockup, tbd what to use finally)

For more than one char, how about something representing formulas (e.g. the sum symbol or this "+-*/" unicode character you use for marking a calculated waypoint)? "..." looks a bit like "to be completed".

> I would also assign the "next" action of the value/formula input field such that the next digit is marked

Perfect.

**Comment by eddiemuc** [comment]
Very good, glad you like it :-) I will start implementing it this way (as time permits...).

> So how does c:geo then know whether its e.g. MinDec or DecDeg?

It doesn't. The current parsing algorithm for plain coordinates is an attempt to parse any input as meaningful as possible (and may we'll be the subject of future discussions, I am aware of this):

Parsing the input from left to right, the Algo tries to find one of the following "token types":
* A hemisphere marker (n, s, e, w or o, Case-sensitive). Each such marker "inverts" the parsed coordinate
* A formula evaluating to a decimal or integer number. If the number is followed by a marker °, ' or " it is interpreted as degree, minute or second. If it is followed by whitespace only then depending on its position in the whole expression ( first. Second or third) it is interpreted as degree. Minute or second

This leads to correct interpretation of all formats I could think of. E.g.

* N48° 12.345
* N48.1234
* 48° 23' 22.345"
* All above variants with formulas instead of numbers


It also leads to interpretation of some false formats as correct I am afraid. If this is seen as a problem we might need to make it stricter. Herr are some examples which would currently be parsed:

* N48.134° S 34" 
* 48 22 44 56 S

> this "+-*/" unicode character

It is not a unicode character but a material.io icon. But we can think of something else

**Comment by ztNFny** [comment]
> It is not a unicode character but a material.io icon. But we can think of something else

how about 
![image](https://user-images.githubusercontent.com/1258173/147975655-371744fe-d568-4047-a100-d1a098c0117c.png)

**Comment by eddiemuc** [comment]
> how about 
> ![image](https://user-images.githubusercontent.com/1258173/147975655-371744fe-d568-4047-a100-d1a098c0117c.png)
> 

Are those Unicode chars or icons? And if the later: where can I find them?

**Comment by ztNFny** [comment]
https://materialdesignicons.com/icon/variable

**Comment by eddiemuc** [comment]
Nightly should contain a first version of the guide now. It has some minor gui hiccups (formula Symbol not correctly centered, border around marked digit button too thin, digit button changes size on tap a bit, hint on "value" test field mising) but should be fully usable otherwise.

Feedback would be highly appreciated whether usage is ok. I don't want to polish the GUI when I have to throw it away later :-)
@Lineflyer

**Comment by eddiemuc** [comment]
GUI stuff should be fixed with PR #12491

**Comment by Lineflyer** [comment]
Just today I did a multi cache which needed one extensive calculation for the final.
I used todays nightly for creating a waypoint for the final (using that nice guided mode) and all went intuitive and as expected. Well done!!

I noticed a smaller UI glitch, wil open a dedicated issue.

IMHO the new waypoint dialog is a real improvement and replacement to the old single waypoint calc. Closing here?!

**Comment by eddiemuc** [comment]
Glad you like it :-)

**Comment by ztNFny** [comment]
Also tested it today and ran into an issue right away: N50°00.A+B+C with A+B+C = 10 results in 50 00.100 instead of 00.010

**Comment by eddiemuc** [comment]
> Also tested it today and ran into an issue right away: N50°00.A+B+C with A+B+C = 10 results in 50 00.100 instead of 00.010

Why would you expect it to result in `00.010`? The formula is not able to guess how many digits after the decimal point you want to have.

However you can tell the formula to reserve overflow space by using the overflow char _ like this: 

`N50 00._(A+B+C)`

This should result in 00.010

**Comment by ztNFny** [comment]
Well it knows that DMM is the only matching coordinate format for that input - just like the cacher manually solving that formula does it.

**Comment by eddiemuc** [comment]
> Well it knows that DMM is the only matching coordinate format for that input - just like the cacher manually solving that formula does it. 

Hm. Current implementation is coordinate-format-agnostic. I personally would prefer this. If we want to go another way then many concepts of the coord parser need to be re-evaluated.

Btw: in current implementation, assuming e.g. `A=2, B=3, C=5`, I would expect that `00.A+B+C` is not evaluated to `0.1` but to `0.2+3+5 = 8.2`. In contrast to `00.(A+B+C)` which should indeed be evaluated to `0.1`. (and `00._(A+B+C)` to `0.01`) There might be a bug here anyway.

**Comment by ztNFny** [comment]
As your implementation is inspired by GCBuddy i checked that one and it calculates the coords i expected

**Comment by eddiemuc** [comment]
> 
> ![Screenshot_20220109-193504](https://user-images.githubusercontent.com/1258173/148695753-252c0db3-1a97-4d34-949f-bcbca7ef9d1c.jpg)
> As your implementation is inspired by GCBuddy i checked that one and it calculates the coords i expected

Yes. The way GCBuddy parses coordinates is actually an aspect of it I don't personally like so much. But this can all be discussed of course

**Comment by ztNFny** [comment]
At least it should be more apparent (but i have no good idea on the how) when such situations occur. In input dialog the coordinate is displayed as calculated (0.26), but in waypoints list only as 0.260 and as i had entered the formula first, then filled only the variables from global vars tab i never noticed the discrepancy - and walked several hundred meter to the wrong location.

**Comment by eddiemuc** [comment]
In yesterday's dev call we discussed and decided a solution to this. I will open a new issue for it to track implementation.
